gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+- The existing health check is also launched locally (the `kubelet` uses the `livenessProbe`) and the situation could be different for pods accessing it remotely. 
+ 
+## Motivation and Scope
+
+
+This project idea aims to get around limitations on Kubernetes’ health check and build an application that: 
+
+
+- Checks CoreDNS health externally through UDP (DNS), from a remote Golang application. 
+- Restart CoreDNS pods by interacting with Kubernetes API through the Golang application, if the response from the cluster and pod IPs is unsatisfactory.  
+
+Thus, making the state of CoreDNS available externally and reliably is important to ensure important services run as they are expected to. A shell script could’ve been used to automate the steps but shell is not good in error handling and may not be reliable depending on which shell (bash/csh/zsh) to use. A Golang program will be much more robust.  
+ 
+### Deliverables  
+
+
+- A standalone Golang application to check CoreDNS health: 
+  - The binary delivered through this project will be a **flexible** application in that it could choose its approach to the problem depending on the environment it is deployed in and the access that has been provided to it.   
+  - The health check program itself doesn’t necessarily reside in the Kubernetes cluster. 
+  - This binary only concerns itself with the health of CoreDNS deployed on Kubernetes. It doesn’t cover other scenarios. 
+
+
+- Proper documentation for the product explaining steps and the concepts used. 
+ 
+ 
+ 
+
+ 
+## Design Proposal 
+
+
+It is necessary to consider all possible cases because the way any team deploys the binary is their **choice** and we should not be forcing a modus operandi to use the product.  
+ 
+Minimal assumptions about the users’ Kubernetes deployment are made and the primary **focus** is to build a program that could easily be adopted by more teams (which may have different preferences of deployment). 
+ 
+In any case, the binary will accomplish the following fundamental tasks: 
+
+
+- Use `kubectl` to extract information about the cluster and its services and pods.  
+ 
+  This will help determine the procedure to follow and also facilitate the subsequent actions to take. This step can be completed only when the application can connect and communicate with the API server. The Go [client](https://github.com/kubernetes/client-go) for Kubernetes has packages that can discover and authenticate with the API server.  
+ 
+- Find the IP address (cluster IP and pod IP) of the CoreDNS service and pods.  
+ 
+  These IP addresses will be used for the `dig` operation in the next steps. It is critical to use both cluster IP and the pod IP. Although it may seem that pinging cluster IP would be sufficient, in real production systems, any point of the system could fail. We cannot make the assumption that Kubernetes networking (forwarding requests from the Service IP to the Pods) will work all the time.  
```

Comment:
> These IP addresses will be used for the `dig` operation in the next steps.

@WJayesh , unnecessary.

> It is critical to use both cluster IP and the pod IP.

@WJayesh , here, you must connect it all ..

```
The previously discovery IP addresses are necessary, because ... 
```

^ you would need to finish it.

Reply:
@WJayesh ,

What happens if CoreDNS service is not found?

What happend if there are no CoreDNS pods?

Reply:
Does this mean a case when CoreDNS isn't installed? I think since most of the clusters should have it by default, those that don't can be shown a message to install it. Is there something more that can be done?